[shopsys] document behavior of less compiling (#646)
shopsys/shopsys: docs/frontend/introduction-to-less.md

> # Indtroduction to LESS

Reviewer: LESS or less or Less?

> # Indtroduction to LESS

Author: I am not sure which one but according their official page less and Less should be fine :)

This document describes behavior of CSS pre-processor [LESS](http://lesscss.org/).

> ## Import files

Reviewer: ```suggestion
## Files import
```

In Shopsys Framework we implement `LESS` by dividing styles into many `LESS components`. Each component has its own file. Filename and component name are the same. We use `@import` command for joining all components contents into one compiled `CSS` file.

Here we can see syntax of `@import` command.
```less
@import 'path/to/directory/component-filename.less';
```

### Usage of @import command

> Best way how to import all related files is to create one file, for example named as `main.less`. This file will contain only `@import` commands. Keep in mind where you place this file. Imported path depends on where this file is placed.

Reviewer: ```suggestion
The best way how to import all related files is to create one file, for example, named `main.less`. This file will contain only `@import` commands. Keep in mind where you place this file. Imported path depends on where this file is placed.
```

#### Unexpected behavior

> There is one thing which you should keep in mind. When you try to import file which does not exist in given file path, compiler will try to find missing file in root directories of files, where is used `@import`.

Reviewer: ```suggestion
There is one thing which you should keep in mind. When you try to import a file which does not exist in the given filepath, the compiler will try to find missing file in root directories of files, where is used `@import`.
```

Reviewer: What do you want to say by "root directories of files, where is used `@import`"? There are more of them? Or only one root directory?

Author: This rule is applied for every file where is used `@import`.

> Let us show this at example. Assuming you have folder structure and files as it is shown below.

Reviewer: ```suggestion
Let us show this at the example. Assuming you have folder structure and files as it is shown below.
```

```
│── root-main.less
│── some-component.less
└─── B
    └── b-main.less
```

```less
/** B/b-main.less */

@import "some-component.less";
```

```less
/* some-component.less */

.some-component {
    color: red;
}
```

```less
/* root-main.less */

@import "B/b-main.less";
```

Result CSS of this example will be.
```less
.some-component {
    color: red;
}
```

> As an explanation of this behavior, given in example above, is that compiler is trying to find a file `some-component.less` firstly in folder where is placed `b-main.less`, then in directory of `root-main.less`. When it could not find required file in any directory, then it will thrown an error during compiling.

Reviewer: There are a few missing articles ("a/an/the") in your document, I suggest using https://app.grammarly.com :slightly_smiling_face:

Author: Thanks for the tip. :)

   └─── layout
   |   └── header.less
   │── helpers.less
   └── main.less
```

For importing of all files in folder `common` there will exist file `main.less` with following code.
```less
/* Import helper classes from current directory */
@import 'helpers.less';

/* Import all global variables from subdirectory core */
@import 'core/variables.less';

/* Import styles defined for header */
@import 'layout/header.less';
```

#### Example 2 - Importing files from another directory
Let us have for this example following folder structure.
```
└─── common
|   └─── core
|   |   └── variables.less
|   └─── layout
|   |   └── header.less
|   │── helpers.less
|   └── main.less
└─── domain2
    └─── core
    |   └── variables.less
    └─── layout
    |   └── footer.less
    └── main.less
```
We can see two files named `main.less` located in two different folders.
- `common/main.less` will import only that type of files which defines common styles for all domains
- `domain2/main.less` will import files which extend, add or modify styles for *domain2*

Now we want to extend styles for *domain2* by changing default colors defined in `core/variables.less` and add styles for footer.

> Code below shows up how would `domain2/main.less` looks like:

Reviewer: ```suggestion
Code below shows up how would `domain2/main.less` look like:
```

# Indtroduction to LESS
This document serves for describing specific behavior of CSS pre-processors LESS.

## Import files
In Shopsys framework we implement LESS by dividing styles into many LESS components. Every component has its own file. File name is same as component name. Getting all these components into one CSS file we accomplish by `@import` command.

In example bellow we can see how looks syntax of this command:

```css
@import 'path/to/directory/component-filename.less';
```

> #### Using @import

Reviewer 2: ```suggestion
### Usage of @import command
```

> Best way to import all related files is create one file, for example named as `main.less`. This file will contain only `@import` commands. Keep in mind where do you place this file. Imported path depends on where this file is placed.

Reviewer 2: ```suggestion
Best way how to import all related files is to create one file, for example named as `main.less`. This file will contain only `@import` commands. Keep in mind where you place this file. Imported path depends on where this file is placed.
```

> ###### Unexcepted behavior

Reviewer 2: ```suggestion
#### Unexpected behavior
```

There is one thing which you should keep in mind. When you try to import file which does not exist in given file path, compiler will try to find missing file in root directories of files, where is used `@import`.

> Let us show this at example. Assuming you have folder structure and files as is shown below:

Reviewer 2: ```suggestion
Let us show this at example. Assuming you have folder structure and files as it is shown below.
```

```
│── root-main.less
│── some-component.less
└─── B
    └── b-main.less
```

> `B/b-main.less`:

Reviewer 2: ```suggestion
```

> ```css

Reviewer 2: ```suggestion
```less
/* B/b-main.less` */
```

@import "some-component.less";
```

> `some-component.less`:

Reviewer 2: ```suggestion
```

> ```css

Reviewer 2: ```suggestion
```less
/* some-component.less */
```

.some-component {
    color: red;
}
```

`root-main.less`:

> ```css

Reviewer 2: ```suggestion
```less
/* root-main.less */
```

@import "B/b-main.less";
```

> Result CSS of this example will be:

Reviewer 2: ```suggestion
Result CSS of this example will be.
```

```css
.some-component {
    color: red;
}
```

> As a explanation of this behavior, given in example above, is that compiler is trying to find file `some-component.less` firstly in folder where is placed `b-main.less`, then in directory of `root-main.less`. When it could not find required file in any directory, then it will thrown an error during compiling.

Reviewer 2: ```suggestion
As an explanation of this behavior, given in example above, is that compiler is trying to find a file `some-component.less` firstly in folder where is placed `b-main.less`, then in directory of `root-main.less`. When it could not find required file in any directory, then it will thrown an error during compiling.
```

> ###### Example 1 - Importing files from current folder and its subfolders

Reviewer 2: ```suggestion
#### Example 1 - Importing files from current folder and its subfolders
```

> Let us have for this example following folder structure:

Reviewer 2: ```suggestion
Let us have for this example following folder structure.
```

```
└── common
   └─── core
   |   └── variables.less
   └─── layout
   |   └── header.less
   │── helpers.less
   └── main.less
```

> For import all files in folder `common` there will be exist file `main.less` with following code:

Reviewer 2: ```suggestion
For importing of all files in folder `common` there will exist file `main.less` with following code.
```

> 

Reviewer 2: ```suggestion
```

> ```css

Reviewer 2: ```suggestion
```less
```

/* Import helper classes from current directory */
@import 'helpers.less';

/* Import all global variables from subdirectory core */
@import 'core/variables.less';

/* Import styles defined for header */
@import 'layout/header.less';
```

> ###### Example 2 - Importing files from another directory

Reviewer 2: ```suggestion
#### Example 2 - Importing files from another directory
```

> Let us have for this example following folder structure:

Reviewer 2: ```suggestion
Let us have for this example following folder structure.
```

```
└─── common
|   └─── core
|   |   └── variables.less
|   └─── layout
|   |   └── header.less
|   │── helpers.less
|   └── main.less
└─── domain2
    └─── core
    |   └── variables.less
    └─── layout
    |   └── footer.less
    └── main.less
```

> We can see two files named `main.less` located in two different folders:

Reviewer 2: ```suggestion
We can see two files named `main.less` located in two different folders.
```

- `common/main.less` will import only that type of files which defines common styles for all domains

> 

Reviewer 2: ```suggestion
```

- `domain2/main.less` will import files which extend, add or modify styles for *domain2*

Now we want to extend styles for *domain2* by changing default colors defined in `core/variables.less` and add styles for footer.
Code below shows up how would `domain2/main.less` looks like:

> ```css

Reviewer 2: ```suggestion
```less
```

# Indtroduction to LESS
This document serves for describing specific behavior of CSS pre-processors LESS.

## Importing LESS files
In Shopsys framework we implement LESS by dividing styles into many LESS components. Every component has its own file. File name is same as component name. Getting all these components into one CSS file we accomplish by `@import` command.

In example bellow we can see how looks syntax of this command:

```css
@import 'path/to/directory/component-filename.less';
```

#### Importing files
Best way to import all related files is create one file, for example named as `main.less`. This file will contain only `@import` commands. Keep in mind where do you place this file. Imported path depends on where this file is placed.

> <!-- `@import` command has set as root directory the directory where the LESS file, using this command, is located. You do not have to even specify current folder by `./`. It is placed before every filepath by default. -->

Reviewer 2: this html comment is not visible in md view mode.
